### PR TITLE
[Feature Store] Fix ingestion with a DropFeatures(entity) step to return an error

### PR DESCRIPTION
**Pull request by tomerm-iguazio**
[ML-3225
](https://jira.iguazeng.com/browse/ML-3225
)
Main changes on api.py in ingest function and on test_spark_engine in  test_ingest_with_steps_drop_features function.

### CODE REVIEW

File: mlrun/feature_store/api.py [hunk 1]
```diff
@@ -405,7 +408,12 @@ def ingest(
         raise mlrun.errors.MLRunInvalidArgumentError(
             "feature set and source must be specified"
         )
-
+    entities_keys = [e.name for e in featureset.spec.entities]
+    for step_name, step_value in featureset.graph.steps.items():
+        if step_name == get_name(name=None, class_name=DropFeatures):
+            dropped_features = step_value.class_args.get("features")
+            if set(dropped_features).intersection(set(entities_keys)):
+                raise mlrun.errors.MLRunInvalidArgumentError(ENTITIES_DROPPED_MESSAGE)
```
Comment (gtopper) [inline mlrun/feature_store/api.py:416]: We don't really do the error message constant thing in mlrun. And if we did, constants are generally not in all caps.

File: mlrun/feature_store/api.py [hunk 2]
```diff
@@ -62,9 +63,11 @@
     run_spark_graph,
 )
 from .retrieval import get_merger, init_feature_vector_graph, run_merge_job
+from .steps import DropFeatures
 
 _v3iofs = None
 spark_transform_handler = "transform"
+ENTITIES_DROPPED_MESSAGE = "Dropping features is illegal on entities keys"
```
Comment (gtopper) [inline mlrun/feature_store/api.py:70]: "DropFeatures can only drop features, not entities"

File: tests/system/feature_store/test_spark_engine.py [hunk 3]
```diff
@@ -1188,6 +1189,25 @@ def test_ingest_with_steps_drop_features(self):
         csv_path_storey = measurements.get_target_path(name="csv")
         read_and_assert(csv_path_spark, csv_path_storey)
 
+        measurements = fstore.FeatureSet(
+            "measurements_spark",
+            entities=[fstore.Entity(key)],
+            timestamp_key="timestamp",
+            engine="spark",
```
Comment (gtopper) [inline tests/system/feature_store/test_spark_engine.py:1196]: The original issue wasn't for spark engine though. Need to add a test for the default engine too.

File: mlrun/feature_store/api.py [hunk 4]
```diff
@@ -405,7 +408,12 @@ def ingest(
         raise mlrun.errors.MLRunInvalidArgumentError(
             "feature set and source must be specified"
         )
-
+    entities_keys = [e.name for e in featureset.spec.entities]
+    for step_name, step_value in featureset.graph.steps.items():
+        if step_name == get_name(name=None, class_name=DropFeatures):
```
Comment (gtopper) [inline mlrun/feature_store/api.py:413]: Doesn't look like `get_name` is doing anything useful here.
```suggestion
        if step_name == "DropFeatures"
```
or
```suggestion
        if step_name == DropFeatures.__name__
```

File: mlrun/feature_store/api.py [hunk 5]
```diff
@@ -405,7 +408,12 @@ def ingest(
...
+            dropped_features = step_value.class_args.get("features")
+            if set(dropped_features).intersection(set(entities_keys)):
```
Comment (gtopper) [inline mlrun/feature_store/api.py:415]: No need to convert to set.
```suggestion
            if set(dropped_features).intersection(entities_keys):
```

File: mlrun/feature_store/api.py [hunk 6]
```diff
@@ -405,7 +408,12 @@ def ingest(
         raise mlrun.errors.MLRunInvalidArgumentError(
             "feature set and source must be specified"
         )
-
+    entities_keys = [e.name for e in featureset.spec.entities]
+    for step_name, step_value in featureset.graph.steps.items():
+        if step_name == get_name(name=None, class_name=DropFeatures):
+            dropped_features = step_value.class_args.get("features")
```
Comment (gtopper) [inline mlrun/feature_store/api.py:414]: We don't want to get `None` if this is missing somehow. We could check for that, but I think this change will suffice:
```suggestion
            dropped_features = step_value.class_args["features"]
```

File: mlrun/feature_store/api.py (hunk 2 above)
Comment (tomerm-iguazio) [reply mlrun/feature_store/api.py:70]: ty!

File: tests/system/feature_store/test_spark_engine.py (hunk 3 above)
Comment (tomerm-iguazio) [reply tests/system/feature_store/test_spark_engine.py:1196]: fixed! in test_feature_store.py - > test_ingest_with_steps_drop_features method

File: mlrun/feature_store/api.py (hunk 5 above)
Comment (tomerm-iguazio) [reply mlrun/feature_store/api.py:415]: fixed! ty!

File: mlrun/feature_store/api.py (hunk 6 above)
Comment (tomerm-iguazio) [reply mlrun/feature_store/api.py]: important comment! ty!

File: mlrun/feature_store/api.py (hunk 1 above)
Comment (tomerm-iguazio) [reply mlrun/feature_store/api.py:416]: fixed!

File: mlrun/feature_store/api.py (hunk 4 above)
Comment (tomerm-iguazio) [reply mlrun/feature_store/api.py]: fixed

File: mlrun/feature_store/api.py [hunk 7]
```diff
@@ -405,7 +417,7 @@ def ingest(
         raise mlrun.errors.MLRunInvalidArgumentError(
             "feature set and source must be specified"
         )
-
+    validate(featureset=featureset)
```
Comment (assaf758) [inline mlrun/feature_store/api.py:420]: function name is too generic, use something like validate_graph_steps()

File: tests/system/feature_store/test_spark_engine.py [hunk 8]
```diff
@@ -1188,6 +1188,28 @@ def test_ingest_with_steps_drop_features(self):
...
+        )
+        measurements.graph.to(DropFeatures(features=[key]))
+        source = ParquetSource("myparquet", path=self.get_remote_pq_source_path())
+        targets = [CSVTarget(name="csv", path=csv_path_spark)]
```
Comment (assaf758) [inline tests/system/feature_store/test_spark_engine.py:1199]: i don't think you need to set targets explicitly for this test, can use the defaults.

**Review by assaf758 — CHANGES_REQUESTED** [review]
Some general comments:
1. The title of PR that fix a bug should start with "Fix" (with capital F) - please rephrase PR tittle to "[Feature Store] Fix ingestion with a DropFeatures of entity step to return an error"
2. change the link to jira to be ML-3225 (and not the explicit link)

File: mlrun/feature_store/api.py [hunk 9]
```diff
@@ -322,6 +323,17 @@ def _rename_source_dataframe_columns(df):
     return df
 
 
+def validate(featureset: Union[FeatureSet, str]):
+    entities_keys = [entity.name for entity in featureset.spec.entities]
+    for step_name, step_value in featureset.graph.steps.items():
+        if step_name == DropFeatures.__name__:
```
Comment (gtopper) [inline mlrun/feature_store/api.py:329]: Can be simplified a bit (no loop):
```suggestion
    if DropFeatures.__name__ in featureset.graph.steps:
        step = featureset.graph.steps[DropFeatures.__name__]
```
Could be even a bit simpler if it were a real dictionary and not an instance of `ObjectDict`.

File: tests/system/feature_store/test_feature_store.py [hunk 10]
```diff
@@ -3874,6 +3874,27 @@ def test_ingest_with_validator_from_uri(self, engine):
...
+            fstore.ingest(
+                measurements,
+                source,
+                infer_options=fstore.InferOptions.schema() + fstore.InferOptions.Stats,
+                run_config=fstore.RunConfig(local=True),
+            )
+            assert (
+                str(ml_run_exception.value)
+                == "DropFeatures can only drop features, not entities"
+            )
```
Comment (gtopper) [inline tests/system/feature_store/test_feature_store.py:3896]: This assert is never run. Needs to be deindented one level.

File: tests/system/feature_store/test_feature_store.py [hunk 11]
```diff
@@ -3874,6 +3874,27 @@ def test_ingest_with_validator_from_uri(self, engine):
...
+        measurements.graph.to(DropFeatures(features=[key]))
+        source = CSVSource(
+            "mycsv", path=os.path.relpath(str(self.assets_path / "testdata.csv"))
+        )
+        with pytest.raises(mlrun.errors.MLRunInvalidArgumentError) as ml_run_exception:
+            fstore.ingest(
+                measurements,
+                source,
+                infer_options=fstore.InferOptions.schema() + fstore.InferOptions.Stats,
+                run_config=fstore.RunConfig(local=True),
```
Comment (gtopper) [inline tests/system/feature_store/test_feature_store.py:3891]: I don't think these parameters are needed for the test.

File: mlrun/feature_store/api.py (hunk 7 above)
Comment (tomerm-iguazio) [reply mlrun/feature_store/api.py:420]: solved

File: tests/system/feature_store/test_spark_engine.py (hunk 8 above)
Comment (tomerm-iguazio) [reply tests/system/feature_store/test_spark_engine.py:1199]: solved

File: mlrun/feature_store/api.py [hunk 12]
```diff
@@ -322,6 +323,17 @@ def _rename_source_dataframe_columns(df):
     return df
 
 
+def validate(featureset: Union[FeatureSet, str]):
```
[the remark this answers is not in the document]
Comment (tomerm-iguazio) [reply mlrun/feature_store/api.py:326]: right fixed.

File: mlrun/feature_store/api.py (hunk 9 above)
Comment (tomerm-iguazio) [reply mlrun/feature_store/api.py:329]: tnx! great idea

File: tests/system/feature_store/test_feature_store.py (hunk 10 above)
Comment (tomerm-iguazio) [reply tests/system/feature_store/test_feature_store.py:3896]: solved

File: tests/system/feature_store/test_feature_store.py (hunk 11 above)
Comment (tomerm-iguazio) [reply tests/system/feature_store/test_feature_store.py:3891]: solved

File: mlrun/feature_store/api.py [hunk 13]
```diff
@@ -322,6 +323,17 @@ def _rename_source_dataframe_columns(df):
     return df
 
 
+def validate_graph_steps(featureset: FeatureSet):
+    if DropFeatures.__name__ in featureset.graph.steps:
+        entities_keys = [entity.name for entity in featureset.spec.entities]
```
Comment (Tankilevitch) [inline mlrun/feature_store/api.py:328]: does this function supposed to be a user facing function, if not lets change it to be a private function.

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Nice to see you already contribute to MLRun! 🥇  
added minor comment.

File: mlrun/feature_store/api.py (hunk 13 above)
Comment (tomerm-iguazio) [reply mlrun/feature_store/api.py]: nice idea! tnx

**Review by Tankilevitch — APPROVED** [review]
LGTM

**Pull request merged**